src/wasm/wat-parser.cpp

```diff
@@ -464,6 +741,39 @@ Result<std::vector<Name>> inlineExports(ParseInput& in) {
   return exports;
 }
 
+// type ::= '(' 'type' id? ft:functype ')' => ft
+template<typename Ctx> MaybeResult<> type(Ctx& ctx, ParseInput& in) {
+  [[maybe_unused]] auto start = in.getPos();
+
+  if (!in.takeSExprStart("type"sv)) {
+    return {};
+  }
+
+  Name name;
+  if (auto id = in.takeID()) {
+    name = *id;
+  }
+
+  // TODO: GC types
+  if (auto type = functype(ctx, in)) {
+    CHECK_ERR(type);
+    if constexpr (parsingTypeDefs<Ctx>) {
+      ctx.builder[ctx.index] = *type;
+    }
+  } else {
+    return in.err("expected type description");
+  }
+
+  if (!in.takeRParen()) {
+    return in.err("expected end of type definition");
+  }
+
+  if constexpr (parsingDecls<Ctx>) {
+    ctx.typeDefs.push_back({name, start});
+  }
```

Comment:
In the future we will have more code containing `if constexpr (parsingTypeDefs<Ctx>)` for struct and array types, so I think the `if constexpr (parsingDecls<Ctx>)` is better off on its own. The other reasonable place to put it would be up at the top right after parsing "(type", but I slightly prefer to record it only at the end once we know there were no errors.